README.md
# WStunnel - Web Sockets Tunnel

- Master:
[![Build Status](https://travis-ci.org/rightscale/wstunnel.svg?branch=master)](https://travis-ci.org/rightscale/wstunnel)
[![Coverage](https://s3.amazonaws.com/rs-code-coverage/wstunnel/cc_badge_master.svg)](https://gocover.io/github.com/rightscale/wstunnel)
- 1.0.7:
[![Build Status](https://travis-ci.org/rightscale/wstunnel.svg?branch=1.0.7)](https://travis-ci.org/rightscale/wstunnel)
[![Coverage](https://s3.amazonaws.com/rs-code-coverage/wstunnel/cc_badge_1.0.7.svg)](https://gocover.io/github.com/rightscale/wstunnel)

WStunnel creates an HTTPS tunnel that can connect servers sitting
behind an HTTP proxy and firewall to clients on the internet. It differs from many other projects
by handling many concurrent tunnels allowing a central client (or set of clients) to make requests
to many servers sitting behind firewalls. Each client/server pair are joined through a rendez-vous token.

At the application level the
situation is as follows, an HTTP client wants to make request to an HTTP server behind a
firewall and the ingress is blocked by a firewall:

```AsciiDoc
  HTTP-client ===> ||firewall|| ===> HTTP-server
```

The WStunnel app implements a tunnel through the firewall. The assumption is that the
WStunnel client app running on the HTTP-server box can make outbound HTTPS requests. In the end
there are 4 components running on 3 servers involved:

- the http-client application on a client box initiates HTTP requests
- the http-server application on a server box behind a firewall handles the HTTP requests
- the WStunnel server application on a 3rd box near the http-client intercepts the http-client's
  requests in order to tunnel them through (it acts as a surrogate "server" to the http-client)
- the WStunnel client application on the server box hands the http requests to the local
  http-server app (it acts as a "client" to the http-server)
The result looks something like this:

```AsciiDoc

HTTP-client ==>\                      /===> HTTP-server
                |                      |
                \----------------------/
            WStunsrv <===tunnel==== WStuncli
```

But this is not the full picture. Many WStunnel clients can connect to the same server and
many http-clients can make requests. The rendez-vous between these is made using secret
tokens that are registered by the WStunnel client. The steps are as follows:

- WStunnel client is initialized with a token, which typically is a sizeable random string,
  and the hostname of the WStunnel server to connect to
- WStunnel client connects to the WStunnel server using WSS or HTTPS and verifies the
  hostname-certificate match
- WStunnel client announces its token to the WStunnel server
- HTTP-client makes an HTTP request to WStunnel server with a std URI and a header
  containing the secret token
- WStunnel server forwards the request through the tunnel to WStunnel client
- WStunnel client receives the request and issues the request to the local server
- WStunnel client receives the HTTP response and forwards that back through the tunnel, where
  WStunnel server receives it and hands it back to HTTP-client on the still-open original
  HTTP request

In addition to the above functionality, wstunnel does some queuing in
order to handle situations where the tunnel is momentarily not open. However, during such
queing any HTTP connections to the HTTP-server/client remain open, i.e., they are not
made aware of the queueing happening.

The implementation of the actual tunnel is intended to support two methods (but only the
first is currently implemented).  
The preferred high performance method is websockets: the WStunnel client opens a secure
websockets connection to WStunnel server using the HTTP CONNECT proxy traversal connection
upgrade if necessary and the two ends use this connection as a persistent bi-directional
tunnel.  
The second (Not yet implemented!) lower performance method is to use HTTPS long-poll where the WStunnel client
makes requests to the server to shuffle data back and forth in the request and response
bodies of these requests.

## Getting Started

You will want to have 3 machines handy (although you could run everything on one machine to
try it out):

- `www.example.com` will be behind a firewall running a simple web site on port 80
- `wstun.example.com` will be outside the firewall running the tunnel server
- `client.example.com` will be outside the firewall wanting to make HTTP requests to
  `www.example.com` through the tunnel

### Download

Release branches are named '1.N.M' and a '1.N' package is created with each revision
as a form of 'latest'.  Download the latest [Linux binary](https://binaries.rightscale.com/rsbin/wstunnel/1.0/wstunnel-linux-amd64.tgz)
and extract the binary. To compile for OS-X or Linux ARM clone the github repo and run
`make depend; make` (this is not tested).

### Set-up tunnel server

On `wstun.example.com` start WStunnel server (I'll pick a port other than 80 for sake of example)

```bash
$ ./wstunnel srv -port 8080 &
2014/01/19 09:51:31 Listening on port 8080
$ curl http://localhost:8080/_health_check
WSTUNSRV RUNNING
$
```

### Start tunnel

On `www.example.com` verify that you can access the local web site:

```bash
$ curl http://localhost/some/web/page
<html> .......
```

Now set-up the tunnel:

```bash
$ ./wstunnel cli -tunnel ws://wstun.example.com:8080 -server http://localhost -token 'my_b!g_$secret!!'
2014/01/19 09:54:51 Opening ws://wstun.example.com/_tunnel
```

### Make a request through the tunnel

On `client.example.com` use curl to make a request to the web server running on `www.example.com`:

```bash

$ curl 'https://wstun.example.com:8080/_token/my_b!g_$secret!!/some/web/page'
<html> .......
$ curl '-HX-Token:my_b!g_$secret!!' https://wstun.example.com:8080/some/web/page
<html> .......
```

### Targeting multiple web servers

The above example tells WStunnel client to only forward requests to `http://localhost`. It is possible to allow the wstunnel to target multiple hosts too. For this purpose the original HTTP client must pass an X-Host header to name the host and WStunnel client must be configured with a regexp that limits the destination web server hostnames it allows. For example, to allow access to `*.some.example.com` over https use:

- `wstunnel cli -regexp 'https://.*\.some\.example\.com' -server https://default.some.example.com ...`
- `curl '-HX-Host: https://www.some.example.com'`

Or to allow access to www.example.com and blog.example.com over http you might use:

- `wstunnel cli -regexp 'http://(www\.example\.com|blog\.example\.com)' -server http://www.example.com ...`
- `curl '-HX-Host: http://blog.example.com'`

Note the use of -server and -regexp, this is because the server named in -server is used when there is no X-Host header. The host in the -server option does not have to match the regexp but it is recommended for it match.

### Using a Proxy

WStunnel client may use a proxy as long as that proxy supports HTTPS CONNECT. Basic authentication may be used if the username and password are embedded in the url. For example, `-proxy http://myuser:mypass@proxy-server.com:3128`. In addition, the command line client will also respect the https_proxy/http_proxy environment variables if they're set. As websocket connections are very long lived, please set read timeouts on your proxy as high as possible.

### Using Secure Web Sockets (SSL)

WStunnel does not support SSL natively (although that would not be a big change). The recommended
approach for using WSS (web sockets through SSL) is to use nginx, which uses the well-hardened
openssl library, whereas WStunnel would be using the non-hardened Go SSL implementation.
In order to connect to a secure tunnel server from WStunnel client use the `wss` URL scheme, e.g.
`wss://wstun.example.com`.
Here is a sample nginx configuration:

````nginx

server {
  listen       443;
  server_name  wstunnel.test.rightscale.com;

  ssl_certificate        <path to crt>;
  ssl_certificate_key    <path to key>;

  # needed for HTTPS
  proxy_set_header X-Forwarded-Proto https;
  proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header Host $http_host;
  proxy_set_header X-Forwarded-Host $host;
  proxy_redirect off;
  proxy_max_temp_file_size 0;

  #configure ssl
  ssl on;
  ssl_protocols SSLv3 TLSv1;
  ssl_ciphers HIGH:!ADH;
  ssl_prefer_server_ciphers on; # don't trust the client
  # caches 10 MB of SSL sessions in memory, faster than OpenSSL's cache:
  ssl_session_cache shared:SSL:10m;
  # cache the SSL sessions for 5 minutes, just as long as today's browsers
  ssl_session_timeout 5m;

  location / {
    root /mnt/nginx;

    proxy_redirect     off;
    proxy_http_version 1.1;

    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $http_host;
    proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
    proxy_buffering off;

    proxy_pass http://127.0.0.1:8080;   # assume wstunsrv runs on port 8080
  }
}
````

### Reading wstunnel server logs

Sample:

```text
Apr  4 17:40:29 srv1 wstunsrv[7808]: INFO HTTP RCV      pkg=WStunsrv token=tech_x... id=19 verb=GET url=/ addr="10.210.2.11, 53.5.22.247" x-host= try=
Apr  4 17:40:29 srv1 wstunsrv[7808]: INFO WS   SND      pkg=WStunsrv token=tech_x... id=19 info="GET /"
Apr  4 17:40:29 srv1 wstunsrv[7808]: INFO WS   RCV      token=tech_x... id=19 ws=0xc20a416d20 len=393
Apr  4 17:40:29 srv1 wstunsrv[7808]: INFO HTTP RET      pkg=WStunsrv token=tech_x... id=19 status=401
```

The first line says that wstunsrv received an HTTP request to be tunneled and assigned it id 19.
The second line says that wstunsrv sent the request onto the appropriate websocket ("WS") to wstuncli.
The third line says that it has received a response to request 19 over the websocket.
The fourth line says that wstunsrv sent an HTTP response back to request 19, and that the status is a 401. What may not be obvious is that because it went round-trip to wstuncli the status code comes from wstuncli.

## Release instructions

- Run the following bash commands.

```bash
make depend
git tag -a $VERSION
git push --tags
```

- Create a branch for the changelog.
- Create the changelog:

```bash
git-chglog -o CHANGELOG.md
git add CHANGELOG.md
```

- Update Readme to reflect new `$VERSION`.
- Commit and push README and CHANGLOG Changes.


CHANGELOG.md
# ChangeLog

## [1.0.7](https://github.com/rightscale/wstunnel/compare/1.0.6...1.0.7)

### Add

* Add Support for Wstuncli port range ([#35](https://github.com/rightscale/wstunnel/issues/35))
* Add support for binding to the specific host ([#23](https://github.com/rightscale/wstunnel/issues/23)) ([#36](https://github.com/rightscale/wstunnel/issues/36))
* Add windows support.
* Add ginkgo binary as a dependency
* Add comments to readme

### Build

* Build and Upload, fixes [#34](https://github.com/rightscale/wstunnel/issues/34) ([#38](https://github.com/rightscale/wstunnel/issues/38))

### Correct

* Correct protocol for tunserv

### Fix

* Fix Travis CI harder
* Fix Makefile to work under Windows/Cygwin
* Fix typo in README
* Fix minor documentation bugs

### Make

* Make GoLint Happy ([#37](https://github.com/rightscale/wstunnel/issues/37))
* Make Travis CI happy as well as the Windows build
* Make version.go editing work for Travis

### Merge

* Merge branch 'master' into patch-1

### Removing

* Removing the reference to HTTPS not supported.

### Switch

* Switch from Godep to dep

### Updated

* updated readme for latest version

### Pull Requests

* Merge pull request [#21](https://github.com/rightscale/wstunnel/issues/21) from chribben/patch-1
* Merge pull request [#29](https://github.com/rightscale/wstunnel/issues/29) from rightscale/ZD156280-1
* Merge pull request [#28](https://github.com/rightscale/wstunnel/issues/28) from rightscale/ph-update-dependencies
* Merge pull request [#25](https://github.com/rightscale/wstunnel/issues/25) from flaccid/patch-1

## [1.0.6](https://github.com/rightscale/wstunnel/compare/1.0.5...1.0.6)

### Add

* add tests to large requests and responses

### Dump

* dump req/resp in case of error, try 2
* dump req/resp in case of error

### Fix

* fix tests not to print to stderr
* fix race condition when reading long messages from websocket in wstuncli
* fix new-websocket race in wstuncli

### Try

* try to fix test race condition [#2](https://github.com/rightscale/wstunnel/issues/2)
* try to fix test race condition

### Tweak

* tweak debug output

## [1.0.5](https://github.com/rightscale/wstunnel/compare/1.0.4...1.0.5)

### Added

* added logging to track ws connect/disconnect

### Fix

* fix error propagation in wsReader
* fix request timeout calculation
* fix panics when pings fail

### Logging

* logging tweak

### Try

* try to fix random test failures

### Tweak

* tweak readme

### Pull Requests

* Merge pull request [#16](https://github.com/rightscale/wstunnel/issues/16) from rightscale/IV-2077_proxy

## [1.0.4](https://github.com/rightscale/wstunnel/compare/1.0.3...1.0.4)

### Attempt

* attempt to test websocket reconnection

### Fix

* fix FD leak in wstuncli; add test

### Update

* update readme

## [1.0.3](https://github.com/rightscale/wstunnel/compare/1.0.2...1.0.3)

### Fix

* fix non-remove requests on tunnel abort; fix non-reopening of WS at client

## [1.0.2](https://github.com/rightscale/wstunnel/compare/1.0.1...1.0.2)

### Added

* added statusfile option

### Readme

* readme fix

### Tweak

* tweak readme

## [1.0.1](https://github.com/rightscale/wstunnel/compare/1.0.0...1.0.1)

### Add

* add more debug info to x-host match failure

### Better

* better non-existent tunnel handling

### Change

* change syslog to LogFormatter, which isn't great either

### Fix

* fix recursive code coverage

### Makefile

* makefile fix 4
* makefile fix 3
* makefile fix 2
* makefile fix

### Support

* support syslog on client; improve syslog log format

### Updated

* updated readme

## 1.0.0

### Acu152721

* acu152721 use exec su, since setuid is forced on pre-start
* acu152721 Have upstart use the www-data user

### Add

* add godeps

### Added

* added upstart config files to tgz
* added support for internal servers
* added more tests
* added log15 dependency, added missing test file
* added upload to S3 to makefile, removed binaries from git
* added whois lookup to tunnel endpoint IP addresses
* added stats request; added explicit GC
* added health check; fixed wstunsrv upstart config
* added upstart conf for wstunsrv and minor edit to wstuncli.conf
* added ubuntu upstart and config, license, and binaries

### Created

* created tunnel package

### Delete

* Delete README.md

### First

* first passing test
* first working version

### Fix

* fix cmdline parsing
* fix async AbortConnection
* fix test suite and coverprofile
* fix host header
* fix x-host error responses; add test cases
* fix code coverage badge
* fix travis.yml
* fix go vet booboo
* fix govet complaints
* fix version
* fix readme
* fix some more logging
* fix some more logging
* fix whois regexp
* fix deadlock in stats handler; add reverse DNS lookup for tunnels
* fix problems with chunked encoding and other transfer encodings
* fix problems with chunked encoding and other transfer encodings
* fix readme headings

### Fixed

* fixed FD leak; added syslog flag; added tunnel keys; sanitize tokens from log
* fixed README instructions

### Godep

* godep hell

### Improved

* improved test for host header
* improved logging; removed read/write timeouts due to bug

### Initial

* Initial commit

### Made

* made x-host header work

### Makefile

* makefile tweak
* makefile tweak
* makefile tweak
* makefile tweak

### Merge

* merge client & server into a single main

### Merge

* Merge branch 'master' of github.com:rightscale/wstunnel
* Merge branch 'master' of github.com:rightscale/wstunnel
* Merge branch 'master' of github.com:rightscale/wstunnel

### More

* More readme changes

### More

* more debugging for wstuncli errors; inadvertant gofmt

### Remove

* remove wstunsrv binary again
* remove read/write timeouts on socket; fix logging typo

### Removed

* Removed logfile and added syslog options

### Removed

* removed default for -server option

### Restrict


tunnel/wstunsrv.go
// Copyright (c) 2014 RightScale, Inc. - see LICENSE

package tunnel

import (
	"bufio"
	"bytes"
	"errors"
	"flag"
	"fmt"
	"io"
	"net"
	"net/http"

	// imported per documentation - https://golang.org/pkg/net/http/pprof/
	_ "net/http/pprof"
	"net/url"
	"os"
	"regexp"
	"runtime"
	"strings"
	"sync"
	"time"

	"github.com/rightscale/wstunnel/whois"
	"gopkg.in/inconshreveable/log15.v2"
)

var _ fmt.Formatter

// The ReadTimeout and WriteTimeout don't actually work in Go
// https://groups.google.com/forum/#!topic/golang-nuts/oBIh_R7-pJQ
//const cliTout = 300 // http read/write/idle timeout

//ErrRetry Error when sending request
var ErrRetry = errors.New("Error sending request, please retry")

const tunnelInactiveKillTimeout = 60 * time.Minute   // close dead tunnels
const tunnelInactiveRefuseTimeout = 10 * time.Minute // refuse requests for dead tunnels

//===== Data Structures =====

const (
	//maxReq max queued requests per remote server
	maxReq = 20
	//minTokenLen min number of chars in a token
	minTokenLen = 16
)

type token string

type responseBuffer struct {
	err      error
	response *bytes.Buffer
}

// A request for a remote server
type remoteRequest struct {
	id         int16               // unique (scope=server) request id
	info       string              // http method + uri for debug/logging
	remoteAddr string              // remote address for debug/logging
	buffer     *bytes.Buffer       // request buffer to send
	replyChan  chan responseBuffer // response that got returned, capacity=1!
	deadline   time.Time           // timeout
	log        log15.Logger
}

// A remote server
type remoteServer struct {
	token           token                    // rendez-vous token for debug/logging
	lastID          int16                    // id of last request
	lastActivity    time.Time                // last activity on tunnel
	remoteAddr      string                   // last remote addr of tunnel (debug)
	remoteName      string                   // reverse DNS resolution of remoteAddr
	remoteWhois     string                   // whois lookup of remoteAddr
	requestQueue    chan *remoteRequest      // queue of requests to be sent
	requestSet      map[int16]*remoteRequest // all requests in queue/flight indexed by ID
	requestSetMutex sync.Mutex
	log             log15.Logger
}

//WSTunnelServer a wstunnel server construct
type WSTunnelServer struct {
	Port                int                     // port to listen on
	Host                string                  // host to listen on
	WSTimeout           time.Duration           // timeout on websockets
	HTTPTimeout         time.Duration           // timeout for HTTP requests
	Log                 log15.Logger            // logger with "pkg=WStunsrv"
	exitChan            chan struct{}           // channel to tell the tunnel goroutines to end
	serverRegistry      map[token]*remoteServer // active remote servers indexed by token
	serverRegistryMutex sync.Mutex              // mutex to protect map
}

// name Lookups
var whoToken string                      // token for the whois service
var dnsCache = make(map[string]string)   // ip_address -> reverse DNS lookup
var whoisCache = make(map[string]string) // ip_address -> whois lookup
var cacheMutex sync.Mutex

func ipAddrLookup(log log15.Logger, ipAddr string) (dns, who string) {
	cacheMutex.Lock()
	defer cacheMutex.Unlock()
	dns, ok := dnsCache[ipAddr]
	if !ok {
		names, _ := net.LookupAddr(ipAddr)
		dns = strings.Join(names, ",")
		dnsCache[ipAddr] = dns
		log.Info("DNS lookup", "addr", ipAddr, "dns", dns)
	}
	// whois lookup
	who, ok = whoisCache[ipAddr]
	if !ok && whoToken != "" {
		who = whois.Whois(ipAddr, whoToken)
		whoisCache[ipAddr] = who
	}
	return
}

//===== Main =====

//NewWSTunnelServer function to create wstunnel from cli
func NewWSTunnelServer(args []string) *WSTunnelServer {
	wstunSrv := WSTunnelServer{}

	var srvFlag = flag.NewFlagSet("server", flag.ExitOnError)
	srvFlag.IntVar(&wstunSrv.Port, "port", 80, "port for http/ws server to listen on")
	srvFlag.StringVar(&wstunSrv.Host, "host", "0.0.0.0", "host for http/ws server to listen on")
	var pidf = srvFlag.String("pidfile", "", "path for pidfile")
	var logf = srvFlag.String("logfile", "", "path for log file")
	var tout = srvFlag.Int("wstimeout", 30, "timeout on websocket in seconds")
	var httpTout = srvFlag.Int("httptimeout", 20*60, "timeout for http requests in seconds")
	var slog = srvFlag.String("syslog", "", "syslog facility to log to")
	var whoTok = srvFlag.String("robowhois", "", "robowhois.com API token")

	srvFlag.Parse(args)

	writePid(*pidf)
	wstunSrv.Log = makeLogger("WStunsrv", *logf, *slog)
	wstunSrv.WSTimeout = calcWsTimeout(*tout)
	whoToken = *whoTok

	wstunSrv.HTTPTimeout = time.Duration(*httpTout) * time.Second
	wstunSrv.Log.Info("Setting remote request timeout", "timeout", wstunSrv.HTTPTimeout)

	wstunSrv.exitChan = make(chan struct{}, 1)

	return &wstunSrv
}

//Start wstunnel server start
func (t *WSTunnelServer) Start(listener net.Listener) {
	t.Log.Info(VV)
	if t.serverRegistry != nil {
		return // already started...
	}
	t.serverRegistry = make(map[token]*remoteServer)
	go t.idleTunnelReaper()

	//===== HTTP Server =====

	var httpServer http.Server

	// Convert a handler that takes a tunnel as first arg to a std http handler
	wrap := func(h func(t *WSTunnelServer, w http.ResponseWriter, r *http.Request)) func(http.ResponseWriter, *http.Request) {
		return func(w http.ResponseWriter, r *http.Request) {
			h(t, w, r)
		}
	}

	// Reqister handlers with default mux
	httpMux := http.NewServeMux()
	httpMux.HandleFunc("/", wrap(payloadHeaderHandler))
	httpMux.HandleFunc("/_token/", wrap(payloadPrefixHandler))
	httpMux.HandleFunc("/_tunnel", wrap(tunnelHandler))
	httpMux.HandleFunc("/_health_check", wrap(checkHandler))
	httpMux.HandleFunc("/_stats", wrap(statsHandler))
	httpServer.Handler = httpMux
	//httpServer.ErrorLog = log15Logger // would like to set this somehow...

	// Read/Write timeouts disabled for now due to bug:
	// https://code.google.com/p/go/issues/detail?id=6410
	// https://groups.google.com/forum/#!topic/golang-nuts/oBIh_R7-pJQ
	//ReadTimeout: time.Duration(cliTout) * time.Second, // read and idle timeout
	//WriteTimeout: time.Duration(cliTout) * time.Second, // timeout while writing response

	// Now create the listener and hook it all up
	if listener == nil {
		t.Log.Info("Listening", "host", t.Host, "port", t.Port)
		laddr := fmt.Sprintf("%s:%d", t.Host, t.Port)
		var err error
		listener, err = net.Listen("tcp", laddr)
		if err != nil {
			t.Log.Crit("Cannot listen", "addr", laddr)
			os.Exit(1)
		}
	} else {
		t.Log.Info("Listener", "addr", listener.Addr().String())
	}
	go func() {
		t.Log.Debug("Server started")
		httpServer.Serve(listener)
		t.Log.Debug("Server ended")
	}()

	go func() {
		<-t.exitChan
		listener.Close()
	}()
}

//Stop wstunnelserver stop
func (t *WSTunnelServer) Stop() {
	t.exitChan <- struct{}{}
}

//===== Handlers =====

// Handler for health check
func checkHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request) {
	fmt.Fprintln(w, "WSTUNSRV RUNNING")
}

// Handler for stats
func statsHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request) {
	// let's start by doing a GC to ensure we reclaim file descriptors (?)
	runtime.GC()

	// make a copy of the set of remoteServers
	t.serverRegistryMutex.Lock()
	rss := make([]*remoteServer, 0, len(t.serverRegistry))
	for _, rs := range t.serverRegistry {
		rss = append(rss, rs)
	}
	// print out the number of tunnels
	fmt.Fprintf(w, "tunnels=%d\n", len(t.serverRegistry))
	t.serverRegistryMutex.Unlock()

	// cut off here if not called from localhost
	addr := r.Header.Get("X-Forwarded-For")
	if addr == "" {
		addr = r.RemoteAddr
	}
	if !strings.HasPrefix(addr, "127.0.0.1") {
		fmt.Fprintln(w, "More stats available when called from localhost...")
		return
	}

	reqPending := 0
	badTunnels := 0
	for i, t := range rss {
		fmt.Fprintf(w, "\ntunnel%02d_token=%s\n", i, cutToken(t.token))
		fmt.Fprintf(w, "tunnel%02d_req_pending=%d\n", i, len(t.requestSet))
		reqPending += len(t.requestSet)
		fmt.Fprintf(w, "tunnel%02d_tun_addr=%s\n", i, t.remoteAddr)
		if t.remoteName != "" {
			fmt.Fprintf(w, "tunnel%02d_tun_dns=%s\n", i, t.remoteName)
		}
		if t.remoteWhois != "" {
			fmt.Fprintf(w, "tunnel%02d_tun_whois=%s\n", i, t.remoteWhois)
		}
		if t.lastActivity.IsZero() {
			fmt.Fprintf(w, "tunnel%02d_idle_secs=NaN\n", i)
			badTunnels++
		} else {
			fmt.Fprintf(w, "tunnel%02d_idle_secs=%.1f\n", i,
				time.Since(t.lastActivity).Seconds())
			if time.Since(t.lastActivity).Seconds() > 60 {
				badTunnels++
			}
		}
		if len(t.requestSet) > 0 {
			t.requestSetMutex.Lock()
			if r, ok := t.requestSet[t.lastID]; ok {
				fmt.Fprintf(w, "tunnel%02d_cli_addr=%s\n", i, r.remoteAddr)
			}
			t.requestSetMutex.Unlock()
		}
	}
	fmt.Fprintln(w, "")
	fmt.Fprintf(w, "req_pending=%d\n", reqPending)
	fmt.Fprintf(w, "dead_tunnels=%d\n", badTunnels)
}

// payloadHeaderHandler handles payload requests with the tunnel token in the Host header.
// Payload requests are requests that are to be forwarded through the tunnel.
func payloadHeaderHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request) {
	tok := r.Header.Get("X-Token")
	if tok == "" {
		t.Log.Info("HTTP Missing X-Token header", "req", r)
		http.Error(w, "Missing X-Token header", 400)
		return
	}
	payloadHandler(t, w, r, token(tok))
}

// Regexp for extracting the tunnel token from the URI
var matchToken = regexp.MustCompile("^/_token/([^/]+)(/.*)")

// payloadPrefixHandler handles payload requests with the tunnel token in a URI prefix.
// Payload requests are requests that are to be forwarded through the tunnel.
func payloadPrefixHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request) {
	reqURL := r.URL.String()
	m := matchToken.FindStringSubmatch(reqURL)
	if len(m) != 3 {
		t.Log.Info("HTTP Missing token or URI", "url", reqURL)
		http.Error(w, "Missing token in URI", 400)
		return
	}
	r.URL, _ = url.Parse(m[2])
	payloadHandler(t, w, r, token(m[1]))
}

// payloadHandler is called by payloadHeaderHandler and payloadPrefixHandler to do the real work.
func payloadHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request, tok token) {
	// create the request object
	req := makeRequest(r, t.HTTPTimeout)
	req.log = t.Log.New("token", cutToken(tok))
	//req.token = tok
	//log_token := cutToken(tok)

	req.remoteAddr = r.Header.Get("X-Forwarded-For")
	if req.remoteAddr == "" {
		req.remoteAddr = r.RemoteAddr
	}

	// repeatedly try to get a response
	for tries := 1; tries <= 3; tries++ {
		retry := getResponse(t, req, w, r, tok, tries)
		if !retry {
			return
		}
	}
}

// getResponse adds the request to a remote server and then waits to get a response back, and it
// writes it. It returns true if the whole thing needs to be retried and false if we're done
// sucessfully or not)
func getResponse(t *WSTunnelServer, req *remoteRequest, w http.ResponseWriter, r *http.Request,
	tok token, tries int) (retry bool) {
	retry = false

	// get a hold of the remote server
	rs := t.getRemoteServer(token(tok), false)
	if rs == nil {
		req.log.Info("HTTP RCV", "addr", req.remoteAddr, "status", "404",
			"err", "Tunnel not found")
		http.Error(w, "Tunnel not found (or not seen in a long time)", 404)
		return
	}

	// Ensure we retire the request when we pop out of this function
	defer func() {
		rs.RetireRequest(req)
	}()

	// enqueue request
	err := rs.AddRequest(req)
	if err != nil {
		req.log.Info("HTTP RCV", "addr", req.remoteAddr, "status", "504",
			"err", err.Error())
		http.Error(w, err.Error(), 504)
		return
	}
	try := ""
	if tries > 1 {
		try = fmt.Sprintf("(attempt #%d)", tries)
	}
	req.log.Info("HTTP RCV", "verb", r.Method, "url", r.URL,
		"addr", req.remoteAddr, "x-host", r.Header.Get("X-Host"), "try", try)
	// wait for response
	select {
	case resp := <-req.replyChan:
		// if there's no error just respond
		if resp.err == nil {
			code := writeResponse(w, resp.response)
			req.log.Info("HTTP RET", "status", code)
			return
		}
		// if it's a non-retryable error then write the error
		if resp.err != ErrRetry {
			req.log.Info("HTTP RET",
				"status", "504", "err", resp.err.Error())
			http.Error(w, resp.err.Error(), 504)
		} else {
			// else we're gonna retry
			req.log.Info("WS   retrying", "verb", r.Method, "url", r.URL)
			retry = true
		}
	case <-time.After(t.HTTPTimeout):
		// it timed out...
		req.log.Info("HTTP RET", "status", "504", "err", "Tunnel timeout")
		http.Error(w, "Tunnel timeout", 504)
	}
	return
}

// tunnelHandler handles tunnel establishment requests
func tunnelHandler(t *WSTunnelServer, w http.ResponseWriter, r *http.Request) {
	if r.Method == "GET" {
		wsHandler(t, w, r)
	} else {
		http.Error(w, "Only GET requests are supported", 400)
	}
}

//===== Helpers =====

// Sanitize the token for logging
func cutToken(tok token) string {
	return string(tok)[0:8] + "..."
}

func (t *WSTunnelServer) getRemoteServer(tok token, create bool) *remoteServer {
	t.serverRegistryMutex.Lock()
	defer t.serverRegistryMutex.Unlock()

	// lookup and return existing remote server
	rs, ok := t.serverRegistry[tok]
	if ok || !create { // return null if create flag is not set
		return rs
	}
	// construct new remote server
	rs = &remoteServer{
		token:        tok,
		requestQueue: make(chan *remoteRequest, maxReq),
		requestSet:   make(map[int16]*remoteRequest),
		log:          log15.New("token", cutToken(tok)),
	}
	t.serverRegistry[tok] = rs
	return rs
}

func (rs *remoteServer) AbortRequests() {
	//logToken := cutToken(rs.tok)
	// end any requests that are queued
l:
	for {
		select {
		case req := <-rs.requestQueue:
			err := fmt.Errorf("Tunnel deleted due to inactivity, request cancelled")
			select {
			case req.replyChan <- responseBuffer{err: err}: // non-blocking send
			default:
			}
		default:
			break l
		}
	}
	idle := time.Since(rs.lastActivity).Minutes()
	rs.log.Info("WS tunnel closed", "inactive[min]", idle)
}

func (rs *remoteServer) AddRequest(req *remoteRequest) error {
	rs.requestSetMutex.Lock()
	defer rs.requestSetMutex.Unlock()
	if req.id < 0 {
		rs.lastID = (rs.lastID + 1) % 32000
		req.id = rs.lastID
		req.log = req.log.New("id", req.id)
	}
	rs.requestSet[req.id] = req
	select {
	case rs.requestQueue <- req:
		// enqueued!
		return nil
	default:
		return errors.New("Too many requests in-flight, tunnel broken?")
	}
}

func (rs *remoteServer) RetireRequest(req *remoteRequest) {
	rs.requestSetMutex.Lock()
	defer rs.requestSetMutex.Unlock()
	delete(rs.requestSet, req.id)
	// TODO: should we close the channel? problem is that a concurrent send on it causes a panic
}

func makeRequest(r *http.Request, httpTimeout time.Duration) *remoteRequest {
	buf := &bytes.Buffer{}
	_ = r.Write(buf)
	return &remoteRequest{
		id:        -1,
		info:      r.Method + " " + r.URL.String(),
		buffer:    buf,
		replyChan: make(chan responseBuffer, 10),
		deadline:  time.Now().Add(httpTimeout),
	}

}

// censoredHeaders, these are removed from the response before forwarding
var censoredHeaders = []string{
	"Connection",
	"Keep-Alive",
	"Te", // canonicalized version of "TE"
	"Trailers",
	"Transfer-Encoding",
}

// Write an HTTP response from a byte buffer into a ResponseWriter
func writeResponse(w http.ResponseWriter, buf *bytes.Buffer) int {
	resp, err := http.ReadResponse(bufio.NewReader(buf), nil)
	if err != nil {
		log15.Info("WriteResponse: can't parse incoming response", "err", err)
		w.WriteHeader(506)
		return 506
	}
	for _, h := range censoredHeaders {
		resp.Header.Del(h)
	}
	// write the response
	copyHeader(w.Header(), resp.Header)
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
	return resp.StatusCode
}

// idleTunnelReaper should be run in a goroutine to kill tunnels that are idle for a long time
func (t *WSTunnelServer) idleTunnelReaper() {
	t.Log.Debug("idleTunnelReaper started")
	for {
		t.serverRegistryMutex.Lock()
		for _, rs := range t.serverRegistry {
			if time.Since(rs.lastActivity) > tunnelInactiveKillTimeout {
				rs.log.Warn("Tunnel not seen for a long time, deleting",
					"ago", time.Since(rs.lastActivity))
				// unlink so new tunnels/tokens use a new RemoteServer object
				delete(t.serverRegistry, rs.token)
				go rs.AbortRequests()
			}
		}
		t.serverRegistryMutex.Unlock()
		time.Sleep(time.Minute)
	}
	//t.Log.Debug("idleTunnelReaper ended")
}
